foubus.py

```diff
@@ -42,411 +53,552 @@
+                    )
+                    continue
+
+                departure_dt = datetime.datetime.fromtimestamp(stu.departure.time)
+                tt.loc[mask, ["realtime", "leave_in"]] = [
+                    True,
+                    departure_dt - now,
                 ]
-                if not row.empty:
-                    assert len(row) == 1, row
-                    # print(row)
-                    # print(stu)
-                    if not stu.departure.time:
-                        logging.warning(
-                            "No departure time: trip: %s stop_time_update: %s",
-                            text_format.MessageToString(
-                                entity.trip_update.trip, as_one_line=True
-                            ),
-                            text_format.MessageToString(stu, as_one_line=True),
-                        )
-                    else:
-                        # row.loc[:,'realtime'] = stu.departure.time
-                        tt.loc[
-                            (tt["trip_id"] == entity.trip_update.trip.trip_id)
-                            & (tt["date"] == entity.trip_update.trip.start_date)
-                            & (tt["stop_sequence"] == stu.stop_sequence)
-                            & (tt["stop_id"] == stu.stop_id),
-                            ["realtime", "leave_in"],
-                        ] = [
-                            True,
-                            datetime.datetime.fromtimestamp(stu.departure.time) - now,
-                        ]
-                        print(row)
-                        updates += 1
-
-    logging.info("TripUpdates for us: %d", updates)
+                updates_count += 1
+
+    logging.info("TripUpdates applied to %d rows", updates_count)
     return tt
 
 
-def next_trips(routes, tt, now):
-    tt["next"] = len(tt) * [False]
-    tt["last"] = len(tt) * [False]
-    # add walking time before picking next (might be too late)
+def next_trips(routes: pd.DataFrame, tt: pd.DataFrame,
+               now: datetime.datetime) -> pd.DataFrame:
+    """
+    For each route in 'routes', identify the next upcoming trip(s) from the
+    timetable 'tt' that departs after 'now'. Mark those in the DataFrame as 'next = True'.
+    Also round 'leave_in' down to the nearest minute.
+
+    Returns a DataFrame subset containing only the upcoming 'next' trips.
+    """
+    tt["next"] = False
+    tt["last"] = False
+
+    # Apply walking time: if 'leave_in' < walking time, we can't catch it.
     def _add_walking_time(row):
-        return row["leave_in"] - pd.Timedelta(minutes=STOPS[row["stop_name"]])
+        walking_minutes = STOPS[row["stop_name"]]
+        return row["leave_in"] - pd.Timedelta(minutes=walking_minutes)
 
     tt["leave_in"] = tt.apply(_add_walking_time, axis=1)
+
+    # For each route, pick up to two upcoming trips
     for (route_id, _, trip_label), _ in routes.iterrows():
-        logging.info("= %s =", trip_label)
```

Comment:
Please keep the logging, while it's in development, it's extremely useful for isolating bugs.